lib/rdoc/rubygems_hook.rb

```diff
@@ -1,17 +1,300 @@
+    end
+
+    options.quiet = !Gem.configuration.really_verbose
+
+    @rdoc = new_rdoc
+    @rdoc.options = options
+
+    store = RDoc::Store.new
+    store.encoding = options.encoding
+    store.dry_run  = options.dry_run
+    store.main     = options.main_page
+    store.title    = options.title
+
+    @rdoc.store = store
+
+    say "Parsing documentation for #{@spec.full_name}"
+
+    Dir.chdir @spec.full_gem_path do
+      @rdoc.parse_files options.files
+    end
+
+    document 'ri',       options, @ri_dir if
+      @generate_ri   and (@force or not File.exist? @ri_dir)
+
+    document 'darkfish', options, @rdoc_dir if
+      @generate_rdoc and (@force or not File.exist? @rdoc_dir)
+  end
+
+  ##
+  # #new_rdoc creates a new RDoc instance.  This method is provided only to
+  # make testing easier.
+
+  def new_rdoc # :nodoc:
+    ::RDoc::RDoc.new
+  end
+
+  ##
+  # Is rdoc documentation installed?
+
+  def rdoc_installed?
+    File.exist? @rdoc_dir
+  end
+
+  ##
+  # Removes generated RDoc and ri data
+
+  def remove
+    base_dir = @spec.base_dir
+
+    raise Gem::FilePermissionError, base_dir unless File.writable? base_dir
+
+    FileUtils.rm_rf @rdoc_dir
+    FileUtils.rm_rf @ri_dir
+  end
+
+  ##
+  # Is ri data installed?
+
+  def ri_installed?
+    File.exist? @ri_dir
+  end
+
+  ##
+  # Prepares the spec for documentation generation
+
+  def setup
+    self.class.load_rdoc
+
+    raise Gem::FilePermissionError, @doc_dir if
+      File.exist?(@doc_dir) and not File.writable?(@doc_dir)
+
+    FileUtils.mkdir_p @doc_dir unless File.exist? @doc_dir
+  end
+
+end
+
 # This class is referenced by RubyGems to create documents.
-# Now, methods are moved to rubygems_plugin.rb.
+# All implementations are moved to the above RubyGemsHook.
+#
+# This class does nothing when this RDoc is installed as a normal gem
+# or a bundled gem.
 #
-# When old version RDoc is not used,
-# this class is not used from RubyGems too.
-# Then, remove this class.
+# This class does generate/remove documents for compatibility when
+# this RDoc is installed as a default gem.
 #
+# We can remove this when all maintained RubyGems remove
+# `rubygems/rdoc.rb`.
 module RDoc
   class RubygemsHook
-    def initialize(spec); end
+    def self.default_gem?
+      File.exist?(File.join(__dir__, "..", "rubygems_plugin.rb"))
+    end
+
+    def initialize(spec)
+      @spe = spec
+    end
+
+    def remove
+      # Do nothing if this is NOT a default gem.
+      return unless self.class.default_gem?
+
+      # Remove generate document for compatibility if this is a
+      # default gem.
+      RubyGemsHook.new(@spec).remove
+    end
 
-    def remove; end
+    def self.generation_hook installer, specs
+      # Do nothing if this is NOT a default gem.
+      return unless default_gem?
 
-    def self.generation_hook installer, specs; end
+      # Generate document for compatibility if this is a default gem.
+      RubyGemsHook.generation_hook(installer, specs)
```

Comment:
RubyGemsHook class doesn't have the `generation_hook` method.

```suggestion
      ::RDoc::RubyGemsHook.generate(installer, specs)
```

Reply:
I've applied the suggested change but can we remove the `::RDoc::` part to simplify this code?

Reply:
I made mistake, sorry...
I think `::RDoc` is unnecessary too.

Reply:
Removed!